### PR TITLE
fix nested flows with middleware

### PR DESCRIPTION
**Pull request by davidyxu**
For the issue mentioned in https://github.com/mobxjs/mobx-state-tree/issues/1271

### CODE REVIEW

File: packages/mobx-state-tree/__tests__/core/create-action-tracking-middleware.test.ts [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+            yield self.innerFlow()
+            return true
+        })
+    }))
+
+test("can handle inner flow", async () => {
+    const m = M.create()
+    addMiddleware(m, middleware)
+    await expect(m.innerFlow()).resolves.toBe(true)
+})
```
Comment (xaviergonz) [inline packages/mobx-state-tree/__tests__/core/create-action-tracking-middleware.test.ts]: could you also check that the actionTrackingMiddlewares are properly called in the right order?
Something like having a string array and then each action tracking middleware hook pushes something to that event and doing a expect(events).toMatchInlineSnapshot

same for the other test

Comment (davidyxu) [reply packages/mobx-state-tree/__tests__/core/create-action-tracking-middleware.test.ts]: Working through the suggested test changes and I think it's uncovering some more issues in the middleware from nested flow.

Just to make sure we're on the same page on expectations, given actions that look like the following:
```
innerFlow: flow(function *() {
  console.log("innerFlow - run 1")
  yield new Promise(resolve => setTimeout(resolve))
  console.log("innerFlow - run 1")
})
outerFlow: flow(function *() {
  console.log("outerFlow - run 1")
  yield self. innerFlow();
  console.log("outerFlow - run 2")
});
```

I would expect to see:
```
outerFlow - onStart
outerFlow - onResume
outerFlow - run 1
innerFlow - onStart
innerFlow - onResume
innerFlow - run 1
innerFlow - onSuspend
outerFlow - onSuspend
outerFlow - onResume
innerFlow - onResume
innerFlow - run 2
innerFlow - onSuspend
innerFlow - onSuccess
outerFlow - onSuspend
outerFlow - onResume
outerFlow - run 2
outerFlow - onSuspend
outerFlow - onSuccess
```

If the plan is to deprecate the old middleware API, would you be open to changing createActionTrackingMiddleware2 interface as follows:
```
export interface IActionTrackingMiddleware2Hooks<TEnv> {
    filter?: (call: IActionTrackingMiddleware2Call<TEnv>) => boolean
    onStart: (call: IActionTrackingMiddleware2Call<TEnv>) => void
    onSuspend?: (call: IActionTrackingMiddleware2Call<TEnv>) => void
    onResume?: (call: IActionTrackingMiddleware2Call<TEnv>) => void
    onFinish: (call: IActionTrackingMiddleware2Call<TEnv>, error?: any) => void
}
```
such that onSuspend and onResume are only necessary if a user needs those hooks (but doesn't complicate the usage for the simple cases)?

Comment (xaviergonz) [reply packages/mobx-state-tree/__tests__/core/create-action-tracking-middleware.test.ts]: Now you see why the second one was created :)

At first I was trying to fix the old one, but the old one was never unit tested and it had some kind of "weird" patterns that I didn't dare to fix in order not to break backwards compatibility, therefore the second one was born

About the new hooks, as long as resume and suspend also gets called properly for sync actions (basically resume right after start and suspend right before finish) I don't see why not :)

Comment (davidyxu) [reply packages/mobx-state-tree/__tests__/core/create-action-tracking-middleware.test.ts]: :) Iterating on the new api is definitely feeling more appealing now, I'll take a stab at it sometime this week.

**Comment by davidyxu** [comment]
@xaviergonz 

I started looking getting the onResume and onSuspend hooks in on this fork:
https://github.com/davidyxu/mobx-state-tree/commit/cd1b71c2d9308ac37c1bf6c7f20adad5678b1058

But I'm still seeing a similar issue as with the old API, where the `outerFlow` hooks aren't being called again when the `innerFlow` resumes (which I would expect to happen because in this case we're calling innerFlow within the context of the outerFlow, so the middleware on innerFlow and outerFlow should both be used firing).
```
    outerFlow [ 'outerFlow (27) - onStart',                                                                                                                  
      'outerFlow (27) - onResume',                                                                                                                           
      'outerFlow - run 1',                                                                                                                                   
      'innerFlow (29) - onStart',                                                                                                                            
      'innerFlow (29) - onResume',
      'innerFlow - run 1',
      'innerFlow (29) - onSuspend',
      'outerFlow (27) - onSuspend',
      'innerFlow (29) - onResume',
      'innerFlow - run 2',
      'innerFlow (29) - onSuspend',
      'innerFlow (29) - onFinish (error: false)',
      'outerFlow (27) - onResume',
      'outerFlow - run 2',
      'outerFlow (27) - onSuspend',
      'outerFlow (27) - onFinish (error: false)' ]
```
For this particular case I could fix it by basically iterating over the call chain and digging up runningActions for parents and calling their onResume/onSuspend in the right order.

However, I don't think this approach would work in more complicated cases because to my understanding, the weak map would only contain running actions for the same middleware. If for example outerFlow was decorated with a different middleware (but innerFlow wasn't), then there isn't a way with the data model to correctly fire off the outerFlow's onResume hook.

It's a pretty significant change, but I think solution would be to have a global map of call -> array of hooks.

**Comment by xaviergonz** [comment]
Can you share your unit test code? I might give it a try

**Comment by davidyxu** [comment]
It should be in the link that I shared: https://github.com/davidyxu/mobx-state-tree/commit/cd1b71c2d9308ac37c1bf6c7f20adad5678b1058#diff-c64a56beaa47a06dd719c54b249a5823

Currently it's just doing console.logs instead of actual expectations but should be pretty straight forward to fix that

**Comment by davidyxu** [comment]
@xaviergonz 
Just wanted to check in to see if you had any luck on your end investigating this. Happy to help setup a few more unit tests or try out any alternative approaches.

**Comment by mweststrate** [comment]
@xaviergonz @davidyxu I notice that this one is open for a really long time already :). Any suggestions on how to proceed?

**Comment by davidyxu** [comment]
@mweststrate 
This PR will fix the exception in the linked issue using the original middleware API. I also have a branch which extends the middleware 2 API with onResume/onSuspend flags.

Unfortunately the way in which middlewares fire when combined with flows is a little unexpected, and from my limited understanding of the MST codebase, it would require some refactoring in the middleware system to get into a more .

If other people are running into the same issue with flows and middlewares mentioned below, I think it makes sense to merge this PR:
https://github.com/mobxjs/mobx-state-tree/issues/1271

The issues around when middleware hooks are called during flows could be tracked with a new giithub issue.

**Comment by mweststrate** [comment]
@davidyxu so if I understand correctly from the issue, you were able to fix the problem with then existing middleware api's, by doing some more work, correct?

**Comment by xaviergonz** [comment]
I tried to make it work without tweaking many things but failed. 

To make this properly work we'd need a refactor to the action context internals so that action each contexts include their root context, the async action that spawned the action context and so on. I did it for mobx-keystone, so the idea is sound, but it is no small feat.

Relevant files if interested:
https://github.com/xaviergonz/mobx-keystone/blob/master/packages/lib/src/action/context.ts
https://github.com/xaviergonz/mobx-keystone/blob/master/packages/lib/src/action/modelFlow.ts
https://github.com/xaviergonz/mobx-keystone/blob/master/packages/lib/src/action/wrapInAction.ts
https://github.com/xaviergonz/mobx-keystone/blob/master/packages/lib/src/actionMiddlewares/actionTrackingMiddleware.ts

**Comment by davidyxu** [comment]
@mweststrate  yes, the immediate problem of the exception is fixed in this PR, but @xaviergonz's suggestion of checking the order in which middleware hooks are called revealed that there's deeper issues to address which requires a larger refactor.

**Comment by jamonholmgren** [comment]
Hey folks, I'm taking over MST as primary maintainer. Could I get an update on where this PR is at and whether you think it'll move forward or if we should close it for now? 🤗

**Comment by jamonholmgren** [comment]
Gentle ping @davidyxu  ^^

**Comment by davidyxu** [comment]
I haven't touched MST in quite a while, but as far as I remember, the PR does fix the uncaught errors mentioned in the linked issue. It however does not fix the underlying issues around the ordering of multiple middlewares.

Those underlying issues look fixable but would involve a much larger refactor than I'm comfortable committing to.  

TLDR: This PR is a small improvement to the original issue, but a larger change is necessary to fix everything discussed here. Please feel free to close it if you'd prefer to wait for a more comprehensive fix.

**Comment by jesse-savary** [comment]
@jamonholmgren I feel that fixing a bug is worth it even though certain MST internals **do need** to be reworked to solve greater issues. Thoughts?

**Comment by jamonholmgren** [comment]
> @jamonholmgren I feel that fixing a bug is worth it even though certain MST internals **do need** to be reworked to solve greater issues. Thoughts?

@jesse-savary I'd be willing to review a new PR that fixes those greater issues; I won't be pushing this forward myself, though, as it's still a fairly unusual edge case to run into any of these. Are you looking for something to work on? 😅
